core/src/nucleus/actions/initialize.rs

```diff
@@ -0,0 +1,113 @@
+extern crate futures;
+use action::{Action, ActionWrapper};
+use agent::actions::commit::commit_entry;
+use context::Context;
+use futures::{future, Async, Future};
+use hash_table::sys_entry::ToEntry;
+use holochain_dna::Dna;
+use instance::dispatch_action_and_wait;
+use nucleus::{
+    ribosome::callback::{genesis::genesis, CallbackParams, CallbackResult},
+    state::NucleusStatus,
+};
+use std::{sync::Arc, thread};
+
+/// Initialize Application, Action Creator
+/// This is the high-level initialization function that wraps the whole process of initializing an
+/// instance. It creates both InitApplication and ReturnInitializationResult actions asynchronously.
+///
+/// Returns a future that resolves to an Ok(NucleusStatus) or an Err(String) which carries either
+/// the Dna error or errors from the genesis callback.
+///
+/// Use futures::executor::block_on to wait for an initialized instance.
+pub fn initialize_application(
+    dna: Dna,
+    context: Arc<Context>,
+) -> Box<dyn Future<Item = NucleusStatus, Error = String>> {
+    if context.state().unwrap().nucleus().status != NucleusStatus::New {
+        return Box::new(future::err(
+            "Can't trigger initialization: Nucleus status is not New".to_string(),
+        ));
+    }
+
+    let context_clone = context.clone();
+
+    thread::spawn(move || {
+        let action_wrapper = ActionWrapper::new(Action::InitApplication(dna.clone()));
+        dispatch_action_and_wait(
+            &context_clone.action_channel,
+            &context_clone.observer_channel,
+            action_wrapper.clone(),
+        );
+
+        // Create Commit Action for Genesis Entry
+        commit_entry(
```

Comment:
Do we need to block_on that future? I replaced the code in the reducer that was also just dispatching an action and not waiting for it to be processed..

Reply:
Hm, I think the old code was too permissive. We should actually be prepared here to handle invalid DNA and let the validation fail. Will change that here and add a test for that..

Reply:
Hmm, ToEntry for DNA always creates an entry from a Dna struct. So with our current code, I can't create a Dna object that would not be able to be commited. So that is why this was ok to not wait for the commit.. but I will push the code that does wait, but without a test that tests it..